### ISSUE TITLE
MapView: ordering of layers

### ISSUE THREAD

**Issue by janusw** [Mapsui/Mapsui#1489]
In the Mapsui.UI.Forms.MapView, the ordering of layers is the following:
1. user layers (base maps + overlays)
2. drawables
3. pins
4. callouts
5. mylocation

From my point of view, there are two issues with this:
1. Mylocation is on top of everything else (possibly hiding drawables, pins & callouts, which I find kind of unnatural)
2. The ordering cannot be changed (any more), since the original ordering is restored in `MapView.HandleLayers`.

Before v3, changing the ordering was possible, and in fact I used to do this via `Map.Layers.Move`. Since #1185 was merged, this does not work any more. The ordering is always restored in `MapView.HandleLayers`.

There are two possible solutions that would fix this for me:
1. Change the default ordering so that mylocation is below the drawables, pins & callouts.
2. Have a way to change the default ordering.

Which one would be preferable?

**Comment by charlenni** [comment]
Short explanation, why this happens: MapView is a convenience layer around MapControl. It should mimic the behavior of Xamarin.Forms original map library. This said, I have to create the extra layers before all other layers and check, that they on top. If this isn‘t done, the next layer (I assume a map tile layer) would cover this layers. Mapsui couldn‘t handle this, because it doesn‘t know anything about special layer ordering.

For me, the MyLocation should on top. You look from above on the map. The highest is yourself. The pins lay on the ground. But that‘s only for me. It seems, you see it different.

**Comment by pauldendulk** [comment]
I think we can improve layer ordering before the release of v4 (but not before the first beta, the beta series can be long). The Mapview now has some weird workarounds because Mapsui lacks a better way of ordering the layers. This also affects performance. Here is a proposal: https://github.com/Mapsui/Mapsui/issues/1491

**Comment by janusw** [comment]
> For me, the MyLocation should on top. You look from above on the map. The highest is yourself. The pins lay on the ground. But that‘s only for me. It seems, you see it different.

Ok, so it seems we disagree on the 'correct' ordering (which is not too surprising, I guess 😄). That means we need a way to allow custom ordering ...

**Comment by janusw** [comment]
> I think we can improve layer ordering before the release of v4

I hope so :)

> The Mapview now has some weird workarounds because Mapsui lacks a better way of ordering the layers. This also affects performance. Here is a proposal: #1491

I don't understand what exactly you propose. In principle the LayerCollection provides all necessary means to achieve any layer ordering you want, right? The problem here seems to be: How to deal with the layers in the MapView? And in particular: How to deal with the fact that there are additional 'internal' layers that were not added by the user?

The idea of `MapView.HandleLayersChanged` seems to be to ensure correct ordering of the internal layers, no matter what the user does. Per se this is not such a bad idea. But it introduces the restriction that one can not change the ordering among the internal layers.

I guess one could simply loosen that restriction via a patch like this:

```diff
-------------------------- Mapsui.UI.Forms/MapView.cs --------------------------
index fd21c24d9..0d6c3ec9d 100644
@@ -472,13 +472,15 @@ namespace Mapsui.UI.Forms
 
         private void HandleLayersChanged(object? sender, LayerCollectionChangedEventArgs args)
         {
             var localRemovedLayers = args.RemovedLayers?.ToList() ?? new List<ILayer>();
             var localAddedLayers = args.AddedLayers?.ToList() ?? new List<ILayer>();
+            var movedLayers = args.MovedLayers?.ToList() ?? new List<ILayer>();
 
             if (localRemovedLayers.Contains(MyLocationLayer) || localRemovedLayers.Contains(_mapDrawableLayer) || localRemovedLayers.Contains(_mapPinLayer) || localRemovedLayers.Contains(_mapCalloutLayer) ||
-                localAddedLayers.Contains(MyLocationLayer) || localAddedLayers.Contains(_mapDrawableLayer) || localAddedLayers.Contains(_mapPinLayer) || localAddedLayers.Contains(_mapCalloutLayer))
+                localAddedLayers.Contains(MyLocationLayer) || localAddedLayers.Contains(_mapDrawableLayer) || localAddedLayers.Contains(_mapPinLayer) || localAddedLayers.Contains(_mapCalloutLayer) ||
+                movedLayers.Contains(MyLocationLayer) || movedLayers.Contains(_mapDrawableLayer) || movedLayers.Contains(_mapPinLayer) || movedLayers.Contains(_mapCalloutLayer))
                 return;
 
             // Remove MapView layers
             RemoveLayers();
 
```

Then a user could decide by himself that he wants to move the internal layers into a different ordering (without undoing it automatically).

**Comment by charlenni** [comment]
@janusw The correct question is: Why are this layers not exposed to the user? It is, because if it isn't, the user could not add a map to MapView. Because if the user add e.g a tile layer, this layer would be above all other layers. The user could not change the order of the layer. And MapView relays on the existence of the special layers. So, if we have Map.Layers following the proposal of @pauldendulk (was the origin of this discussion), then we could add the special layers to a high value group and all other layers, the user adds later, are added to the default group.

**Comment by janusw** [comment]
> So, if we have Map.Layers following the proposal of @pauldendulk (was the origin of this discussion), then we could add the special layers to a high value group and all other layers, the user adds later, are added to the default group.

Right. However, one disadvantage of the interface proposed in #1491 is that there is no way to restrict access to a specific group ...

Let's say we put the 'user' layers into group 0 and the 'special/internal' layers of the MapView into group 1. Then, how do you prevent a user from adding a layer to group 1 (or group 2)?

Maybe an alternative would be to use a top-level class `LayerCollection` that would contain one or several `LayerGroup` objects (ordered in a particular way), which in turn contain an ordered list of layers.

Then you could have one group exposed publicly and keep the other one private ...

**Comment by charlenni** [comment]
But @janusw let's be clear: we could not work around each problem. So the user of Mapsui has to think a little.

I wanted to prevent an odd first time experience. If you start with MapView, you will add a map layer, that is under all your pins and geometries. If you start to delete groups of layers or move them around, then you should know, what is going on behind the scenes. We couldn't, and I assume, we wouldn't do that. So grouping layers would be better than now. And you are free to rearrange the layers in any order you like.

**Comment by charlenni** [comment]
Ok, from my side I see no problem. Should work. And if it helps.

**Comment by janusw** [comment]
> Ok, from my side I see no problem. Should work. And if it helps.

Alright, it would definitely help me. I just submitted it as PR #1494.

**Comment by janusw** [comment]
This issue is fixed by #1494 for me. I think it still makes sense to introduce further improvements like layer groups, which should be discussed in #1491.
